File: examples/python/propulsion_pulse.py
"""显式驱动 Apollo 11 RCS 短脉冲和 DPS 可调推力档。"""

import numpy as np

from apollo_sim import (
    ApolloPropulsionPlantFactory,
    ApolloState,
    DpsCommand,
    PropulsionCommand,
    RcsCommand,
    RcsThrusterId,
)


def main() -> None:
    factory = ApolloPropulsionPlantFactory.apollo11_touchdown()
    plant = factory.spawn(ApolloState.identity())

    # 历史标签描述喷管/羽流方向；spec.force_direction_body 才是飞船受力方向。
    thruster = RcsThrusterId.A1F
    short_request = PropulsionCommand(
        rcs=RcsCommand.single_pulse(thruster, 1_000_000),
        dps=DpsCommand.off(),
    )
    rcs_step = plant.step(short_request)
    print(
        f"RCS {thruster.name}: requested="
        f"{rcs_step.requested_command.rcs.on_time_ns[thruster]} ns, "
        f"applied_gate="
        f"{rcs_step.applied.rcs.applied_gate_on_time_ns[thruster]} ns, "
        f"mean_thrust={rcs_step.applied.rcs.mean_thrust_n[thruster]:.3f} N"
    )

    # RCS 有关断尾迹；先显式空跑一 tick，避免下一行的合力混入残余 RCS 冲量。
    plant.step(PropulsionCommand.all_off())

    dps_spec = factory.propulsion_spec.dps
    target_gimbal_x_rad = 2.0 * np.pi / 180.0
    target_gimbal_z_rad = -1.0 * np.pi / 180.0
    dps_step = plant.step(
        PropulsionCommand(
            rcs=RcsCommand.all_off(),
            # 这里提交的是目标摆角；AppliedDps 返回受 GDA 摆速限制后的实际摆角。
            dps=DpsCommand.variable(
                dps_spec.variable_min_thrust_n,
                gimbal_x_rad=target_gimbal_x_rad,
                gimbal_z_rad=target_gimbal_z_rad,
            ),
        )
    )
    print(
        "DPS: "
        f"mode={dps_step.applied.dps.mode.value}, "
        f"thrust={dps_step.applied.dps.thrust_n:.3f} N, "
        f"target_gimbal=({np.rad2deg(target_gimbal_x_rad):.3f}°, "
        f"{np.rad2deg(target_gimbal_z_rad):.3f}°), "
        f"actual_gimbal=({np.rad2deg(dps_step.applied.dps.gimbal_x_rad):.6f}°, "
        f"{np.rad2deg(dps_step.applied.dps.gimbal_z_rad):.6f}°), "
        f"mean_force_body={dps_step.applied.mean_wrench_body.force_body_n}"
    )

    # DPS 命令同样不跨 tick 隐式保持；连续工作必须每周期重复提交。
    plant.step(PropulsionCommand.all_off())


if __name__ == "__main__":
    main()

File: examples/rust/propulsion_pulse.rs
//! Rust 推进接口最小例程：直接操作一枚 RCS 喷口，再操作 DPS。
//!
//! 这里没有控制器或喷口分配器；调用方提交的就是 16 路阀门时间与 DPS 档位。

use apollo_mujoco::{
    ApolloPropulsionPlantFactory, ApolloState, DpsCommand, PropulsionCommand, RcsCommand,
    RcsThrusterId,
};
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let factory = ApolloPropulsionPlantFactory::apollo11_touchdown()?;
    let spec = factory.propulsion_spec();
    let mut plant = factory.spawn(ApolloState::ZERO)?;

    // 请求 7 ms 会按实机最小脉冲约束提升为 14 ms；数组第 0 路是 A1U。
    let thruster_id = RcsThrusterId::new(0).expect("0 is a valid RCS id");
    let pulse = plant.step(PropulsionCommand {
        rcs: RcsCommand::single_pulse(thruster_id, 7_000_000),
        dps: DpsCommand::Off,
    })?;
    println!(
        "RCS {} requested=7ms applied={}ns mean_thrust={:.3}N",
        thruster_id.label(),
        pulse.applied.rcs[thruster_id.index()].applied_gate_on_time_ns,
        pulse.applied.rcs[thruster_id.index()].mean_thrust_n,
    );
    println!("RCS mean wrench: {:?}", pulse.applied.mean_wrench_body);

    // DPS 可调档会把低于 1,050 lbf 的请求钳到该档下限；摆角以弧度输入，
    // 但它是 GDA 的目标位置，不会在一个控制周期内瞬间到达。
    let descent = plant.step(PropulsionCommand {
        rcs: RcsCommand::OFF,
        dps: DpsCommand::Variable {
            thrust_n: 1_000.0,
            gimbal_x_rad: 2.0_f64.to_radians(),
            gimbal_z_rad: -1.0_f64.to_radians(),
        },
    })?;
    println!(
        "DPS mode={:?} applied_thrust={:.3}N actual_gimbal=({:.6}°, {:.6}°) direction={:?}",
        descent.applied.dps.mode,
        descent.applied.dps.thrust_n,
        descent.applied.dps.gimbal_x_rad.to_degrees(),
        descent.applied.dps.gimbal_z_rad.to_degrees(),
        descent.applied.dps.force_direction_body,
    );

    assert_eq!(
        pulse.applied.rcs[thruster_id.index()].applied_gate_on_time_ns,
        spec.rcs_thrusters[thruster_id.index()].minimum_pulse_ns
    );
    assert_eq!(descent.applied.dps.thrust_n, spec.dps.variable_min_thrust_n);
    let actual_gimbal_magnitude = descent
        .applied
        .dps
        .gimbal_x_rad
        .hypot(descent.applied.dps.gimbal_z_rad);
    let maximum_tick_delta = spec.dps.gimbal_rate_rad_s * factory.timing().control_step_seconds();
    assert!((actual_gimbal_magnitude - maximum_tick_delta).abs() < 1.0e-15);
    Ok(())
}
